### PR TITLE
Added 3D Plotting Functionality to plotter.py

### PR DESCRIPTION
**Pull request by PACarniglia**
Adds three-dimensional plotting functionality to the plotter.py function. Plotting of truth data, measurements (including clutter and true detection distinction), and tracks with error bars are supported. The underlying 2D plotting functionality is unchanged. Plotting dimensionality is now selected by the user when instantiating new Plotter objects by passing a plotter.py Dimension Enum object. Currently Dimension.TWO and Dimension.THREE are supported but can be extended in the future.

The legend is now populated with a dictionary where the keys show up as labels and values associated with those keys appear as the handles. This implementation avoids duplicate legend entries if plotting the same data multiple times.

A unit test file, test_plotter.py, was also created to ensure subsequent changes to plotter.py do not break current functionality.

### CODE REVIEW

File: stonesoup/plotter.py [hunk 1]
```diff
@@ -151,36 +189,38 @@ def plot_measurements(self, measurements, mapping, measurement_model=None,
...
-                    **kwargs):
+                    err_freq=1, **kwargs):
         """Plots track(s)
 
-        Plots each track generated, generating a legend automatically. If ``uncertainty=True``,
-        uncertainty ellipses are plotted. If ``particle=True``, particles are plotted.
-        Tracks are plotted as solid lines with point markers and default colors.
-        Uncertainty ellipses are plotted with a default color which is the same for all tracks.
+        Plots each track generated, generating a legend automatically. If ``uncertainty=True``
+        and is being plotted in 2D, error elipses are plotted. If beingp plotted in
```
Comment (nperree-dstl) [inline stonesoup/plotter.py:217]:
```suggestion
        and is being plotted in 2D, error elipses are plotted. If being plotted in
```

File: stonesoup/plotter.py [hunk 2]
```diff
@@ -189,13 +229,17 @@ def plot_tracks(self, tracks, mapping, uncertainty=False, particle=False, track_
             Set of tracks which will be plotted. If not a set, and instead a single
             :class:`~.Track` type, the argument is modified to be a set to allow for iteration.
         mapping: list
-            List of 2 items specifying the mapping of the x and y components of the state space.
+            List of 3 items specifying the mapping of the x, y, and z position
```
Comment (nperree-dstl) [inline stonesoup/plotter.py:232]:
```suggestion
            List of items specifying the mapping of the position
```

File: stonesoup/plotter.py [hunk 3]
```diff
@@ -151,36 +189,38 @@ def plot_measurements(self, measurements, mapping, measurement_model=None,
...
-        Uncertainty ellipses are plotted with a default color which is the same for all tracks.
+        Plots each track generated, generating a legend automatically. If ``uncertainty=True``
+        and is being plotted in 2D, error elipses are plotted. If beingp plotted in
+        3D, uncertainty bars are plotted every :attr:`err_freq` measurement, default
+        plots unceratinty bars at every track step. Tracks are plotted as solid
+        lines with point markers and default colors. Uncertainty bars are plotted
+        with a default color which is the same for all tracks.
 
-        Users can change linestyle, color and marker using keyword arguments. Uncertainty ellipses
+        Users can change linestyle, color and marker using keyword arguments. Uncertainty bars
```
Comment (nperree-dstl) [inline stonesoup/plotter.py:223]:
```suggestion
        Users can change linestyle, color and marker using keyword arguments. Uncertainty ellipses/bars
```

File: stonesoup/plotter.py [hunk 4]
```diff
@@ -189,13 +229,17 @@ def plot_tracks(self, tracks, mapping, uncertainty=False, particle=False, track_
             Set of tracks which will be plotted. If not a set, and instead a single
             :class:`~.Track` type, the argument is modified to be a set to allow for iteration.
         mapping: list
-            List of 2 items specifying the mapping of the x and y components of the state space.
+            List of 3 items specifying the mapping of the x, y, and z position
+            components of the state space.
         uncertainty : bool
-            If True, function plots uncertainty ellipses.
+            If True, function plots uncertainty bars in x, y, and z.
```
Comment (nperree-dstl) [inline stonesoup/plotter.py:235]:
```suggestion
            If True, function plots uncertainty ellipses or bars.
```

**Review by nperree-dstl — COMMENTED** [review]
All looks good to me from a first glance at the code. A couple of documentation things, but really nothing major. I've not tried to use it yet but I will have a go when I get a chance.

File: stonesoup/tests/test_plotter.py [hunk 5]
```diff
@@ -0,0 +1,110 @@
...
+
+from stonesoup.hypothesiser.distance import DistanceHypothesiser
+from stonesoup.measures import Mahalanobis
+
+from stonesoup.dataassociator.neighbour import NearestNeighbour
+from stonesoup.types.state import GaussianState
+
+from stonesoup.types.track import Track
+
+np.random.seed(1991)
```
[the remark this answers is not in the document]
Comment (PACarniglia) [reply stonesoup/tests/test_plotter.py]: I've deleted line 29, the measurements are generated randomly now in-line with the other tests.

**Comment by PACarniglia** [comment]
@sdhiscocks I was told by @ekhunter123 that the remaining failing tests can be ignored, is this still the case for this PR?

**Comment by ekhunter123** [comment]
> @sdhiscocks I was told by @ekhunter123 that the remaining failing tests can be ignored, is this still the case for this PR?

I think here I was referencing the comment on my pull request #587, that CircleCi is having a problem lately with the dependency cache. I told @PACarniglia about this and wondered if the same applies for this PR. Perhaps "ignore" was too strong of a word for me to use. The fact that the tests fail still gives important information and maybe more work can be done to fix it. On the other hand, maybe it is just a problem with CircleCi, I don't know.

**Comment by sdhiscocks** [comment]
> 

You are correct @ekhunter123, the tests are broken for pull requests; the same dependency cache issue. I'll see if I can work out a way to resolve this, but any fix will probably only resolve itself for new PRs.

**Review by sdhiscocks — APPROVED** [review]
Looks good :+1:

Minor suggestion:
```diff
diff --git a/stonesoup/plotter.py b/stonesoup/plotter.py
index 9cdc0c80..999f7566 100644
--- a/stonesoup/plotter.py
+++ b/stonesoup/plotter.py
@@ -53,11 +53,7 @@ class Plotter:
     """
 
     def __init__(self, dimension=Dimension.TWO):
-        if isinstance(dimension, type(Dimension.TWO)):
-            self.dimension = dimension
-        else:
-            raise TypeError("""%s is an unsupported type for \'dimension\';
-                            expected type %s""" % (type(dimension), type(Dimension.TWO)))
+        self.dimension = Dimension(dimension)
         # Generate plot axes
         self.fig = plt.figure(figsize=(10, 6))
         if self.dimension is Dimension.TWO:  # 2D axes
diff --git a/stonesoup/tests/test_plotter.py b/stonesoup/tests/test_plotter.py
index 258dab64..46cd7ed2 100644
--- a/stonesoup/tests/test_plotter.py
+++ b/stonesoup/tests/test_plotter.py
@@ -81,7 +81,7 @@ plotter = Plotter()
 
 
 def test_dimension_raise():
-    with pytest.raises(TypeError):
+    with pytest.raises(ValueError):
         Plotter(dimension=1)  # expected to raise TypeError
 
 
```

Wrapping the input with Enum should allow both `Plotter(dimension=3)` and `Plotter(dimension=Dimension.THREE)` to work.

**Comment by PACarniglia** [comment]
> Looks good 👍
> 
> Minor suggestion:
> 
> ```diff
> diff --git a/stonesoup/plotter.py b/stonesoup/plotter.py
> index 9cdc0c80..999f7566 100644
> --- a/stonesoup/plotter.py
> +++ b/stonesoup/plotter.py
> @@ -53,11 +53,7 @@ class Plotter:
>      """
>  
>      def __init__(self, dimension=Dimension.TWO):
> -        if isinstance(dimension, type(Dimension.TWO)):
> -            self.dimension = dimension
> -        else:
> -            raise TypeError("""%s is an unsupported type for \'dimension\';
> -                            expected type %s""" % (type(dimension), type(Dimension.TWO)))
> +        self.dimension = Dimension(dimension)
>          # Generate plot axes
>          self.fig = plt.figure(figsize=(10, 6))
>          if self.dimension is Dimension.TWO:  # 2D axes
> diff --git a/stonesoup/tests/test_plotter.py b/stonesoup/tests/test_plotter.py
> index 258dab64..46cd7ed2 100644
> --- a/stonesoup/tests/test_plotter.py
> +++ b/stonesoup/tests/test_plotter.py
> @@ -81,7 +81,7 @@ plotter = Plotter()
>  
>  
>  def test_dimension_raise():
> -    with pytest.raises(TypeError):
> +    with pytest.raises(ValueError):
>          Plotter(dimension=1)  # expected to raise TypeError
>  
>  
> ```
> 
> Wrapping the input with Enum should allow both `Plotter(dimension=3)` and `Plotter(dimension=Dimension.THREE)` to work.

@sdhiscocks the reason for switching the current implementation of dimension as an int to dimension as an Enum is twofold; it sanitizes the input as users will have to use a Dimension instance (this was recommended by @ekhunter123), and makes it easier for future developers to extend the plotting capabilities present in Plotter.py without having to be restricted to just adding additional dimensions - in theory, the dimension parameter need not be a spatial number but could rather be sort of a "plot type" attribute.

**Review by jswright-dstl — APPROVED** [review]

**Pull request merged**